python/pyiceberg/catalog/rest.py

```diff
@@ -0,0 +1,373 @@
+    error: ErrorResponseMessage = Field()
+
+
+class RestCatalog(Catalog):
+    token: TokenResponse
+    config: ConfigResponse
+
+    host: str
+
+    def __init__(self, name: str, properties: Properties, host: str, token: str):
+        self.host = host
+        self.token = self._fetch_access_token(token)
+        self.config = self._fetch_config()
+        super().__init__(name, properties)
+
+    def _split_token(self, token: str) -> Tuple[str, str]:
+        """Splits the token in a client id and secret
+
+        Args:
+            token: The token with a semicolon as a separator
+
+        Returns:
+            The client id and secret
+        """
+        client, secret = token.split(":")
+        return client, secret
+
+    @property
+    def headers(self) -> Properties:
+        return {AUTHORIZATION_HEADER: f"{BEARER_PREFIX} {self.token.access_token}", "Content-type": "application/json"}
+
+    def url(self, endpoint: str, prefixed: bool = True, **kwargs) -> str:
+        """Constructs the endpoint
+
+        Args:
+            prefixed: If the prefix return by the config needs to be appended
+
+        Returns:
+            The base url of the rest catalog
+        """
+
+        url = self.host
+        url = url if url.endswith("/") else url + "/"
+
+        if prefixed:
+            url += self.config.overrides.get(PREFIX, "")
+            url = url if url.endswith("/") else url + "/"
+
+        return url + endpoint.format(**kwargs)
+
+    def _fetch_access_token(self, token: str) -> TokenResponse:
+        client, secret = self._split_token(token)
+        data = {GRANT_TYPE: CLIENT_CREDENTIALS, CLIENT_ID: client, CLIENT_SECRET: secret, SCOPE: CATALOG_SCOPE}
+        response = requests.post(self.url(Endpoints.get_token, prefixed=False), data=data)
+        try:
+            response.raise_for_status()
+        except HTTPError as exc:
+            self._handle_non_200_response(exc, {401: BadCredentialsError})
+
+        return TokenResponse(**response.json())
+
+    def _fetch_config(self) -> ConfigResponse:
+        response = requests.get(self.url(Endpoints.get_config, prefixed=False), headers=self.headers)
+        response.raise_for_status()
+        return ConfigResponse(**response.json())
+
+    def _split_namespace_and_table(self, identifier: Union[str, Identifier]) -> Properties:
+        identifier = self.identifier_to_tuple(identifier)
+        return {"namespace": NAMESPACE_SEPARATOR.join(identifier[:-1]), "table": identifier[-1]}
+
+    def _split_namespace_and_name(self, identifier: Union[str, Identifier]) -> Dict[str, Union[Tuple[str, ...], str]]:
+        identifier = self.identifier_to_tuple(identifier)
+        return {"namespace": identifier[:-1], "name": identifier[-1]}
+
+    def _handle_non_200_response(self, exc: HTTPError, error_handler: Dict[int, Type[Exception]]):
+        try:
+            response = ErrorResponse(**exc.response.json())
+        except Exception:  # pylint: disable=W0012,W0703
```

Comment:
`request.json()` says that it will raise `requests.exceptions.JSONDecodeError`. Is it a good idea to catch `Exception` instead of the specific one?

Reply:
Yes, I was also thinking of that. However, since this is already a sad path, I figured to catch everything. I don't have a very strong opinion on this, but when another error bubbles up, it might not be caught.

Reply:
I've changed this to `JSONDecodeError`